File: algorithm study/BOJ/2. Silver/16507.js
// boj 16507 어두운 건 무서워
// prefix sum

const input = require("fs").readFileSync("./example.txt").toString().split("\n");
// const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

const [r, c, q] = input[0].split(" ").map(Number);

let arr = [];
for (let i = 1; i < r + 1; i++) arr.push(input[i].split(" ").map(Number));
let temp = Array.from(Array(r + 1), () => new Array(c + 1).fill(0));

for (let i = 1; i <= r; i++) temp[i][1] = arr[i - 1][0];
//행별 누적합이 아닌 좌표별 누적합
for (let i = 1; i < r + 1; i++) {
  for (let j = 1; j < c + 1; j++) {
    temp[i][j] = arr[i - 1][j - 1] + temp[i][j - 1] + temp[i - 1][j] - temp[i - 1][j - 1];
  }
}

for (let i = 0; i < q; i++) {
  let [x1, y1, x2, y2] = input[r + i + 1].split(" ").map(Number);

  //겹치는 부분 빼기, 두 번 뺀 부분 더하기
  let sum = temp[x2][y2] - temp[x1 - 1][y2] - temp[x2][y1 - 1] + temp[x1 - 1][y1 - 1];
  let div = (x2 - x1 + 1) * (y2 - y1 + 1);

  console.log(Math.floor(sum / div));
}

File: algorithm study/BOJ/2. Silver/16507.cpp
// boj 16507 어두운 건 무서워
// prefix sum
#include <iostream>
using namespace std;

int R, C, Q;  // 1 ≤ R, C ≤ 1000, 1 ≤ Q ≤ 10000
int arr[1001][1001];
int temp[1001][1001];  // 누적합

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> R >> C >> Q;
    for (int i = 1; i <= R; i++) {
        for (int j = 1; j <= C; j++) {
            int K;  // 1 ≤ K ≤ 1,000
            cin >> K;
            arr[i][j] = K;
        }
    }

    for (int i = 1; i <= R; i++) temp[i][1] = arr[i][1];  // 행별 누적합
    for (int i = 1; i <= R; i++) {
        for (int j = 2; j <= C; j++) temp[i][j] = temp[i][j - 1] + arr[i][j];
    }

    for (int i = 0; i < Q; i++) {
        int sum = 0;
        int r1, c1, r2, c2;  // 1 ≤ r1 ≤ r2 ≤ R, 1 ≤ c1 ≤ c2 ≤ C
        cin >> r1 >> c1 >> r2 >> c2;

        if (r1 == r2 & c1 == c2)
            sum = arr[r1][c1];
        else {
            for (int a = r1; a <= r2; a++) {
                sum += temp[a][c2] - temp[a][c1 - 1];
            }
        }
        cout << sum / ((r2 - r1 + 1) * (c2 - c1 + 1)) << "\n";
    }
    return 0;
}